device.cpp

```diff
@@ -11,4 +11,110 @@ Flashcart::Flashcart(const char* name, const size_t max_length)
         flashcart_list = new std::vector<Flashcart*>();
     }
     flashcart_list->push_back(this);
-}
\ No newline at end of file
+}
+
+uint32_t Flashcart::readFlash(uint32_t address, uint32_t length, uint8_t *buffer) {
+    if (length <= 0) return 0;
+    setWriteState(false);
+
+    uint32_t curpos;
+    for (curpos = 0; curpos < length;) {
+        uint32_t read_sz = rawRead(address + curpos, length - curpos, buffer + curpos);
+        if (!read_sz) break;
+
+        curpos += read_sz;
+        showProgress(curpos, length, "Reading");
```

Comment:
`length-1` otherwise we get shit like 99/100

Reply:
either length-1 or curpos+1. Makes sense when you think about it, because one of them is 0-based and the other is 1-based. Need to compensate.

Reply:
I also need to move it in front of the action, not leave it in back.